cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java

```diff
@@ -0,0 +1,445 @@
+    private final Integer numPods;
+    private final Supplier<BackOff> backoffSupplier;
+    protected String namespace;
+    private final AdminClientProvider adminClientProvider;
+
+    KafkaRoller(Vertx vertx, PodOperator podOperations,
+                long pollingIntervalMs, long operationTimeoutMs, Supplier<BackOff> backOffSupplier,
+                StatefulSet ss, Secret clusterCaCertSecret, Secret coKeySecret) {
+        this(vertx, podOperations, pollingIntervalMs, operationTimeoutMs, backOffSupplier,
+                ss, clusterCaCertSecret, coKeySecret, new DefaultAdminClientProvider());
+    }
+
+    KafkaRoller(Vertx vertx, PodOperator podOperations,
+                long pollingIntervalMs, long operationTimeoutMs, Supplier<BackOff> backOffSupplier,
+                StatefulSet ss, Secret clusterCaCertSecret, Secret coKeySecret,
+                AdminClientProvider adminClientProvider) {
+        this.namespace = ss.getMetadata().getNamespace();
+        this.cluster = Labels.cluster(ss);
+        this.numPods = ss.getSpec().getReplicas();
+        this.backoffSupplier = backOffSupplier;
+        this.clusterCaCertSecret = clusterCaCertSecret;
+        this.coKeySecret = coKeySecret;
+        this.vertx = vertx;
+        this.operationTimeoutMs = operationTimeoutMs;
+        this.podOperations = podOperations;
+        this.pollingIntervalMs = pollingIntervalMs;
+        this.adminClientProvider = adminClientProvider;
+    }
+
+    /**
+     * Returns a Future which completed with the actual pod corresponding to the abstract representation
+     * of the given {@code pod}.
+     */
+    protected Future<Pod> pod(Integer podId) {
+        return podOperations.getAsync(namespace, KafkaCluster.kafkaPodName(cluster, podId));
+    }
+
+    private final ScheduledExecutorService singleExecutor = Executors.newSingleThreadScheduledExecutor(
+        runnable -> new Thread(runnable, "kafka-roller"));
+
+    private ConcurrentHashMap<Integer, RestartContext> podToContext = new ConcurrentHashMap<>();
+    private Predicate<Pod> podNeedsRestart;
+
+    /**
+     * Asynchronously perform a rolling restart of some subset of the pods,
+     * completing the returned Future when rolling is complete.
+     * Which pods get rolled is determined by {@code podNeedsRestart}.
+     * The pods may not be rolled in id order, due to the {@linkplain KafkaRoller rolling algorithm}.
+     * @param podNeedsRestart Predicate for determining whether a pod should be rolled.
+     * @return A Future completed when rolling is complete.
+     */
+    Future<Void> rollingRestart(Predicate<Pod> podNeedsRestart) {
+        this.podNeedsRestart = podNeedsRestart;
+        List<Future> futures = new ArrayList<>(numPods);
+        List<Integer> podIds = new ArrayList<>(numPods);
+        for (int podId = 0; podId < numPods; podId++) {
+            podIds.add(podOperations.isReady(namespace, podName(podId)) ? podIds.size() : 0, podId);
```

Comment:
as offline chat I would add a comment about the ordering and that the rolling is always first about not-ready pods.